pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -144,6 +237,102 @@ impl VectoredReadBuilder {
     }
 }
 
+pub(crate) struct ChunkedVectoredReadBuilderInner {
+    /// Start block number
+    start_blk_no: usize,
+    /// End block number (exclusive).
+    end_blk_no: usize,
+    /// Blobs (metadata, end offset) ordered by blobs start offset.
+    blobs_at: VecMap<u64, (u64, BlobMeta)>,
+    max_read_size: Option<usize>,
+    chunk_size: usize,
+}
+
+/// Computes x / d rounded up.
+fn div_round_up(x: usize, d: usize) -> usize {
+    (x + (d - 1)) / d
+}
+
+impl ChunkedVectoredReadBuilderInner {
+    /// Start building a new vectored read.
+    ///
+    /// Note that by design, this does not check against reading more than `max_read_size` to
+    /// support reading larger blobs than the configuration value. The builder will be single use
+    /// however after that.
+    pub(crate) fn new(
+        start_offset: u64,
+        end_offset: u64,
+        meta: BlobMeta,
+        max_read_size: Option<usize>,
+        chunk_size: usize,
+    ) -> Self {
+        let mut blobs_at = VecMap::default();
+        blobs_at
+            .append(start_offset, (end_offset, meta))
+            .expect("First insertion always succeeds");
+
+        let start_blk_no = start_offset as usize / chunk_size;
+        let end_blk_no = div_round_up(end_offset as usize, chunk_size);
+        Self {
+            start_blk_no,
+            end_blk_no,
+            blobs_at,
+            max_read_size,
+            chunk_size,
+        }
+    }
+
+    /// Returns the end offset of the last blob.
+    pub(crate) fn last_end_offset(&self) -> u64 {
+        // SAFETY(unwrap): `self.blobs_at` always has a least one entry since construction.
+        let (_, (end, _)) = self.blobs_at.as_slice().last().unwrap();
+        *end
+    }
+
+    /// Attempts to extend the current read with a new blob if:
+    ///
+    /// - The start offset matches with the current end of the vectored read
+    /// - The end block number matches the current end block number.
+    ///
+    /// The resulting size also must be below the max read size.
+    pub(crate) fn extend(&mut self, start: u64, end: u64, meta: BlobMeta) -> VectoredReadExtended {
+        tracing::trace!(start, end, "trying to extend");
+        let start_blk_no = start as usize / self.chunk_size;
+        let end_blk_no = div_round_up(end as usize, self.chunk_size);
+        if (self.last_end_offset() == start || self.end_blk_no == start_blk_no + 1) && {
```

Comment:
```
2 blocks of size 100
         100    199
|------|  |------|
0      99
   A.        B.
```

The `end_blk_no` is exclusive, which means that we have (start_blk_no=0, end_blk_no=1) for both blocks. In this case, the condition `A.end_blk_no == B.start_blk_no + 1` is satisfied.

I think `end_blk_no` is confusing. I will find a better name.

Reply:
> The end_blk_no is exclusive

Makes sense.

> which means that we have (start_blk_no=0, end_blk_no=1) for both blocks

I don't see this, but you're right they would be merged.
For A: `start_blk_no = 10 / 100 = 0` and `end_block_no = (119 + 99) / 100 = 2`.
For B: `start_blk_no = 120 / 100 = 1` and `end_block_no = (160 + 99) / 100 = 2`.
 
---

Let's take a different example.

```
0     99  100   199 200   299
|------|  |------| |------|
   a          b      c d
```

Blob [a, b]: `start_blk_no = 0` `end_blk_no = 2`
Blob [c, d]: `start_blk_no = 2` `end_blk_no = 3`

In this case it would be desirable to merge but we don't.

Reply:
15fa58c02852f7bf652a2f0ae380c8857a10f3b4 should take care of this.